### ISSUE TITLE
Fontawesome or Glyphicon Icons available to set for page icons and module heading.

### ISSUE THREAD

**Issue by thabaum** [dnnsoftware/Dnn.Platform#2703]
Just an idea to make it available to set an icon from fontawesome or glyphicon for a page icon and maybe even the title of modules.

If a user has them available in the theme settings by adding the line needed to the site skin theme to make these available then they can possibly select an option to use a an icon from one of these free icon sites.

Text field to simply add the "fas fa-comment" to a textbox after selecting the radio button to use available free web icon sets instead of an image or file to select an icon to help represent a page.

Also might be nice to be able to set a url for the icon to use for a page.  But I really like the idea of using these freely available icons as a quick way to help represent page content.

**Comment by OllyHodgson** [comment]
I'm a bit torn with this. I love the idea in general, but it's not a simple one. 

Firstly, there are accessibility concerns which will complicate it a bit. Like any other image, the content editor would need to provide appropriate alternative text _or_ hide it from accessible technology (e.g. screen readers). e.g. if it's purely a decorative icon (and/or there's some descriptive text alongside it), include `aria-hidden="true"`. In other scenarios it's not quite as straightforward - but `aria-label="Whatever this icon means"` is often the best way. Of course, `aria-label` is usually only visible to screen-readers, not visual browsers... 

A bit more on the subject:
- https://fontawesome.com/how-to-use/on-the-web/other-topics/accessibility
- https://getbootstrap.com/docs/3.4/components/#glyphicons-how-to-use

Secondly, the world seems to be moving away from icon fonts, and towards inline SVG elements. FontAwesome 5 offers a way to inline the SVGs using JS - but it's not open source like FA4 was. Iconic offers a similar thing.

Finally, there are so many icon sets out there, and there's very little consistency between implementations (i.e. each one has it's own unique way of adding it to the page). In my experience, no two designers will choose the same icon set for a given project. So if we choose an icon set to include in Platform, we've already chosen the wrong one. 

Perhaps this should sit in the realm of extensions / plugins? And perhaps this conversation (RFC?) should be about providing a really good way to package an SVG or raster icon library for DNN, and how that might integrate into the CKEditor, the inline editor, and any other editor providers?

**Comment by OllyHodgson** [comment]
Also: I seem to remember note there is already an icon set included with DNN (or maybe it's in Evoq? You can choose it in site settings) so we might be partway there already. IIRC there's no obvious way to use those icons in page content though.

**Comment by mitchelsellers** [comment]
This is something that I'll work to try and add as a discussion point to one of the TAG meetings.  I agree with @OllyHodgson that there are a number of considerations that can impact this.  Another one to think about is the licensing requirements associated with the inclusion of a third-party as well as impacts to skins etc.

**Comment by valadas** [comment]
I totally agree with all of @OllyHodgson comments. 

The very old icons set is here https://github.com/dnnsoftware/Dnn.Platform/tree/development/Website/Icons/Sigma
Many extensions use that in their manifest for the module icon (hardoced path to the Sigma Set)
Many css files of Dnn controls (file upload, dropdowns, date pickers, etc) also refer to some of those icons in a harcoded path way.

Now the icon set to use is configurable in the site settings (why this got migrated into the persona bar and so many other things did not make it into the persona bar escapes me, I never heard of anyone using a different icon set). But anyway, people could add a folder with another name and change that setting and get all new icons (except where they are hardcoded for Sigma)

There is also an API (DotNetNuke.Entities.Icons.IconController) to find the path to an icon given a name, size and style (ie: edit, 16, standard would return /Icons/configured-icon-set/edit_16x16_standard.png

Now for the persona bar, it is a totally different story, there are svg icons in a folder, those dont really need a size and style since they are basically just a shape...  Those also have an 'api' since they are part of the common bundle for persona bar extensions.

So we currently have 2 different icon sets both of which are some kind of a breaking change to remove.

As for Fontawesome, I think the free set is ok to use according to this license here  https://fontawesome.com/license/free and I _think_ it is the most used icons (popularity)

So, I don't know what the best course of action would be, but I would hate to have a third new set of icons. Also, I think all design and styles should be the responsability of the theme. So I have mixed opinions on the whole idea. But definitelly worth thinking about. I will post back if I come up with an incredible idea...

**Comment by thabaum** [comment]
I just think it would be nice to be able to click in the CK Editor and have the line auto-inserted for the icon.  Even if this was an option in source mode.  To be able to enter <i class="fas fa-"></i> inline for example would be something simple and time saving for someone developing.  Maybe some type of built in clipboard that remembers simple code that you use a lot and being able to search in from within the source editor would save time.  I suppose CK editor in source would be nice to have some type of IntelliSense built into it as well.

As for persona bar being able to access these icons it would be great as well for page icons and for module title icons if there was an option for such a thing.

**Comment by OllyHodgson** [comment]
That feels like something a CKEditor plugin ought to handle. I seem to remember it's perfectly possible to install CKEditor plugins in DNN, but not exactly super easy.

**Comment by valadas** [comment]
Yeah, it is possible and not that easy, but you can look at https://github.com/DNN-Connect/CKEditorProvider/pull/84 I made one for quickly inserting html5 video tags and it is still not perfect and never got time to return to it. But if it gives you some pointers on how to do it...

The other problem is that if you want to have those html tags work, then you need to load the iconset css and the webfont, those add quite some weight. My preference would be to have individual files (svg ideally) and just insert them like any image.

**Comment by thabaum** [comment]
Thanks for the tip, I might be able to build something for it with a reference to work with.

Also icons do not show up in the WYSIWYG editor, it would be nice to actuallly see what you get in the editor.  However I did open this as Page and Module so I suppose I went sideways on topic.  Thinking back to when I posted this I was thinking more about being able to set fontawesome icons up for pages / modules to show in menu's and title's.  However it would be more user friendly if it had a way to not only set the icons for pages mainly, but to also allow you to keep up with the current release of font awesome as they continuously add more free open source icons to use.  

Then also theme and module creators would have this built into DNN as an option that if this is turned on then all of these icons would be available with a way to browse and select them :)  The module creator just needs to say in the module requirements this feature needs to be turned on.  End result would be possibly faster development of modules that have a consistent look and feel.
